### PR TITLE
Optimizations to both CFG construction and resulting CFG

### PR DESCRIPTION
**Pull request by DarkDimius**
### Motivation
(based on #1042)
Pay-server spends ~15% of time in functions affected in this PR.
I'd expect that new implementation is faster, but I'm yet to validate that.
I also find it easier to understand and reason about(thought this might be because I wrote it now, and old one was written by me ~2 years ago).

### Test plan
Existing test + manual check that all updates make "better" cfgs.

See included automated tests.

### CODE REVIEW

File: cfg/CFG.cc [hunk 1]
```diff
@@ -28,50 +28,54 @@ CFG::CFG() {
 }
 
 CFG::ReadsAndWrites CFG::findAllReadsAndWrites(core::Context ctx) {
-    UnorderedMap<core::LocalVariable, UnorderedSet<BasicBlock *>> reads;
```
Comment (DarkDimius) [inline cfg/CFG.cc]: Indexing by LocalVariable was, to put it nicely, a _very_ bad idea.

File: test/testdata/cfg/math.rb.cfg-json.exp [hunk 2]
```diff
@@ -626,34 +626,6 @@
       }
      }
     },
-    {
-     "bind": {
-      "unique_name": "\u003cself\u003e",
```
Comment (DarkDimius) [inline test/testdata/cfg/math.rb.cfg-json.exp]: this is correct, we can now prove this variable to be unused

File: cfg/builder/builder_finalize.cc [hunk 3]
```diff
@@ -239,8 +239,25 @@ void CFGBuilder::removeDeadAssigns(core::Context ctx, const CFG::ReadsAndWrites
                 continue;
             }
 
-            auto fnd = RnW.reads.find(bind.bind.variable);
```
Comment (DarkDimius) [inline cfg/builder/builder_finalize.cc]: this is the actual change in behavior.
It used to see if the variable is "ever" read.
Now instead, we see if the variable is read in _this_ block or is consumed by the outgoing blocks.

**Comment by DarkDimius** [comment]
Though it's passing tests here, it's failing on pay-server.
We're lacking a test to reproduce.

**Comment by DarkDimius** [comment]
> We're lacking a test to reproduce.

No, we had a test. I just didn't notice that changes to test/testdata/cfg/side_effects.rb.cfg.exp were wrong.

**Comment by DarkDimius** [comment]
I've checked that performance on pay-server is fine after this change.

File: cfg/builder/builder_finalize.cc [hunk 4]
```diff
@@ -239,8 +239,25 @@ void CFGBuilder::removeDeadAssigns(core::Context ctx, const CFG::ReadsAndWrites
...
+            }
+            if (!wasRead) {
+                for (const auto &arg : it->bexit.elseb->args) {
+                    if (arg.variable == bind.bind.variable) {
+                        wasRead = true;
+                        break;
+                    }
+                }
+            }
+            if (!wasRead) {
```
Comment (jez) [inline cfg/builder/builder_finalize.cc]: Would this code read simpler if it was written like

```cpp
if (wasRead) {
  continue;
}
```

Comment (jez) [reply cfg/builder/builder_finalize.cc]: Ah jk i think it might be hard because we don't always do `expIt++`.

File: cfg/builder/builder_entry.cc [hunk 5]
```diff
@@ -79,8 +79,8 @@ unique_ptr<CFG> CFGBuilder::buildFor(core::Context ctx, ast::MethodDef &md) {
     dealias(ctx, *res);
     CFG::ReadsAndWrites RnW = res->findAllReadsAndWrites(ctx);
     computeMinMaxLoops(ctx, RnW, *res);
-    removeDeadAssigns(ctx, RnW, *res);
     fillInBlockArguments(ctx, RnW, *res);
+    removeDeadAssigns(ctx, RnW, *res); // requires block arguments to be filled
```
Comment (jez) [inline cfg/builder/builder_entry.cc]: small note, it took me a second to realize that this meant "require basic block arguments to be filled"

after looking at the implementation of `removeDeadAssigns` it's clear that what this means, but we might want to rename `fillInBlockArguments` to `fillInBasicBlockArguments`.

Comment (jez) [reply cfg/builder/builder_entry.cc]: (non blocking)

File: cfg/builder/builder_finalize.cc [hunk 6]
```diff
@@ -305,106 +320,84 @@ void CFGBuilder::fillInBlockArguments(core::Context ctx, CFG::ReadsAndWrites &Rn
...
+    const vector<UnorderedSet<core::LocalVariable>> &writesByBlock = RnW.writes;
+    const vector<UnorderedSet<core::LocalVariable>> &deadByBlock = RnW.dead;
 
     // iterate ver basic blocks in reverse and found upper bounds on what could a block need.
     vector<UnorderedSet<core::LocalVariable>> upperBounds1(cfg.maxBasicBlockId);
     bool changed = true;
-
-    while (changed) {
-        changed = false;
+    {
```
Comment (jez) [inline cfg/builder/builder_finalize.cc]: Why use scope blocks in this function? Are you trying to trigger destructors, or are they just there for grouping?

I usually prefer dividing sections with comments, but if this is just your personal style that's fine too.

**Pull request merged**